packages/world/src/modules/keysintable/getKeysInTable.sol

```diff
@@ -11,34 +13,69 @@ import { KeysInTable } from "./tables/KeysInTable.sol";
  * Note: this util can only be called within the context of a Store (e.g. from a System or Module).
  * For usage outside of a Store, use the overload that takes an explicit store argument.
  */
-function getKeysInTable(bytes32 tableId) view returns (bytes32[][] memory keysInTable) {
-  // Get the keys with the given table
-
+function getKeysInTable(bytes32 tableId) view returns (bytes32[][] memory keyTuples) {
   /**
-   * Note: this module does not yet support composite keys.
-   * Therefore we create an array of key tuples but only fill the first entry.
+   * Note: this module only supports up to 5 composite keys.
    */
-  bytes32[] memory keysInTableRaw = KeysInTable.getKeys(tableId);
 
-  keysInTable = new bytes32[][](keysInTableRaw.length);
+  Schema schema = StoreSwitch.getKeySchema(tableId);
+
+  uint256 numFields = schema.numFields();
+  uint256 length = KeysInTable.lengthKeys0(tableId);
+  keyTuples = new bytes32[][](length);
 
-  for (uint256 i; i < keysInTableRaw.length; i++) {
-    keysInTable[i] = new bytes32[](1);
-    keysInTable[i][0] = keysInTableRaw[i];
+  for (uint256 i; i < length; i++) {
+    keyTuples[i] = new bytes32[](numFields); // the length of the key tuple depends on the schema
+
+    if (numFields > 0) {
+      keyTuples[i][0] = KeysInTable.getItemKeys0(tableId, i);
+      if (numFields > 1) {
+        keyTuples[i][1] = KeysInTable.getItemKeys1(tableId, i);
+        if (numFields > 2) {
+          keyTuples[i][2] = KeysInTable.getItemKeys2(tableId, i);
+          if (numFields > 3) {
+            keyTuples[i][3] = KeysInTable.getItemKeys3(tableId, i);
+            if (numFields > 4) {
+              keyTuples[i][4] = KeysInTable.getItemKeys4(tableId, i);
+            }
+          }
+        }
+      }
+    }
```

Comment:
since we're interested in the full array, but just need to change the format of the result data, it might be cheaper to first load the full arrays and then change the format (instead of going though the table library for every item, which has some overhead)

Reply:
not blocking though, we can add an issue to track and come back to gas optimizing this

Reply:
True. Either way the gas impact this has had is pretty crazy. I am on the side of leaving it for now and refactoring this whole module in future to support any number of keys, codegen etc.